### PR TITLE
Stack vs sample proposal and outer loop

### PR DESCRIPTION
**Pull request by elsony**
Signed-off-by: eyuen <eyuen@redhat.com>

### What does this PR do?
Move design doc from gdoc (https://docs.google.com/document/d/1tbzZK9aWtpw5AtoZmUl1Jy8x8w9N5hhjXu3c26VYiYY/edit?usp=sharing and https://docs.google.com/document/d/1SL3ZJyAy8lp973EzQpY36VjTJHoyGXl2zUXokPiXwK8/edit?usp=sharing) to design doc in repo for public access.

### What issues does this PR fix or reference?
Design doc for https://github.com/devfile/api/issues/279 and https://github.com/devfile/api/issues/346

### Is your PR tested? Consider putting some instruction how to test your changes
Design doc changes only

#### Docs PR
Design doc changes only

### CODE REVIEW

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+            buildContext: ${PROJECTS_ROOT}/build
+            location: Dockerfile
+            args: [ "arg1", "arg2", "arg3" ]
+            rootRequired: false
+
+`imageName`: name of the generated image (or we may use an id for referencing the image). This imageName can also be fully qualified to allow pushing of the image to a specific image registry.
+
+`buildContext`: Path of source directory to establish build context.  Default to ${PROJECT_ROOT} (optional)
+
+`location`: Dockerfile location which can be an URL or a path relative to buildContext
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:56]: Should this default to `Dockerfile` to replicate the semantics of `docker build ${buildContext}`?

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+`args`: Argument list for the docker build (optional)
+
+`rootRequired`: Specifies whether a privileged builder pod is required.  Default is false. (optional)
+
+__Note:__
+1. A common pattern will be using a global variable to define the image name (`myimage` in the example above) so that it can be easily referred to in the deploy step later.
+2. The `apply` command with the `image` component is optional. If the `apply` command that references an image component does not exist, then the image build will be done at the startup automatically (similar to the behaviour of the existing `kubernetes` components).  If there is an `apply` command that references the `image` component, then the user will need to create a composite command to chain the image build apply command with the deploy command (see the deployment command section) to complete the build and deploy.
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:64]: Which `apply` command is this referring to?

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+### Example using a file Dockerfile with registry and secret for the image push:
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:66]:
```suggestion
### Example using a Dockerfile with registry and secret for the image push:
```

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+__Alternative approach:__ we can consider reusing the `exec` type except that these build types are special and we may not want to incorporate the full image build command into an `exec` command. Therefore, I chose to use a different type of command here.
+
+### Example using a file Dockerfile as part of the app:
+    variables:
+      myimage: myimagename
+    components:
+    - name: mydockerfileimage
+        image:
+        imageName: {{myimage}}
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:45]:
```suggestion
    - name: mydockerfileimage
      image:
        imageName: {{myimage}}
```

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+### Example using a Dockerfile stored in a git repo with push registry without apply command:
+    variables:
+    myimage: myimagename
+    components:
+    - name: mydockerfileimage
+        image:
+        imageName: {{myimage}}
+        dockerfile:
+            git:
+            remotes:
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:116]:
```suggestion
    - name: mydockerfileimage
      image:
        imageName: {{myimage}}
        dockerfile:
          git:
            remotes:
```

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+                origin: "https://github.com/odo-devfiles/nodejs-ex.git"
+            location: Dockerfile
+            args: [ ]
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:119]: What does `.components[].image.dockerfile` look like in totality? I'm seeing multiple conflicting sets of properties (`buildContext`/`location` vs `id` vs `git`)

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+
+The git definition will be the same as the one in `starterProjects` definition that supports `checkoutFrom`
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:121]: Can we just reuse `projects`/`starterProjects` here?

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+
+`location`: the location of the dockerfile within the repo. If `checkoutFrom` is being used, the location will be relative to the `checkoutFrom`.
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:123]: This seems unclear to me -- `checkoutFrom` refers to a remote, so what does "relative" mean in this context?

File: docs/proposals/devfile/outer-loop-build-and-deploy.md [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+    selector:
+        matchLabels:
+        app: {{.COMPONENT_NAME}}
+    template:
+        metadata:
+        creationTimestamp: null
+        labels:
+            app: {{.COMPONENT_NAME}}
+        spec:
+        containers:
```
Comment (amisevsk) [inline docs/proposals/devfile/outer-loop-build-and-deploy.md:255]: Could you fix the indentation errors throughout -- I don't know if it's a misplaced tab character or something.

File: docs/proposals/registry/sample-vs-stack.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
+# Proposal for sample vs stack in a devfile registry
+
+## Overview
+
+The existing devfile registry focuses on the stack support to provide generic language/framework/runtime, e.g. Node, Java maven, Quarkus, etc.,  support to build and run user applications. These devfiles are called ___*stack devfiles*___.  There is another kind of devfile in a devfile registry that is tailored for a building and running a specific application. These devfiles are referred to as ___*sample devfiles*___ (example: https://github.com/redhat-developer/devfile-sample).
```
Comment (amisevsk) [inline docs/proposals/registry/sample-vs-stack.md:5]: I think e.g. `***sample devfiles***` looks more markdown-y and renders the same as `___*sample devfiles*___`

File: docs/proposals/registry/sample-vs-stack.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    ]
+
+The bold entries highlighted the differences between a stack entry vs a sample entry in the devfile index registry. 
+
+## Registry library change
+The registry library and the registry REST API will provide a filtering mechanism to specify which type (sample or stack) is needed in the query.
+
+
+
+
```
Comment (amisevsk) [inline docs/proposals/registry/sample-vs-stack.md:195]: Trim trailing blank lines
```suggestion

```

File: docs/proposals/registry/sample-vs-stack.md [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+            "language": "nodejs",
+            "git": {
+                "remotes": {
+                    "origin": "https://github.com/redhat-developer/devfile-sample/"
+                }
+            }
+        }
+    ]
+
+The bold entries highlighted the differences between a stack entry vs a sample entry in the devfile index registry.
```
Comment (GeekArthur) [inline docs/proposals/registry/sample-vs-stack.md:188]: The `bold entries` are not displayed after moving docs from google docs, we need manually added them again by using markdown.

**Review by GeekArthur — CHANGES_REQUESTED** [review]

File: docs/proposals/devfile/outer-loop-build-and-deploy.md (hunk 7 above)
Comment (elsony) [reply docs/proposals/devfile/outer-loop-build-and-deploy.md]: We are just sharing the git definition with starterProjects/projects. This section has nothing to do with starter project so it does not apply.

File: docs/proposals/devfile/outer-loop-build-and-deploy.md (hunk 1 above)
Comment (elsony) [reply docs/proposals/devfile/outer-loop-build-and-deploy.md]: When running `docker build` command, it either uses the current working directory or the build context specified when running the command. When tools are trying to execute the dockerfile, we need to define the default directory that the tools can run the docker build command. With the described behaviour, it means the tools can find the Dockerfile in the project root by default which is the most common setup of the project root.

**Pull request closed without merging**